### PR TITLE
This integrates Mozilla's Speech API in the gateway frontend

### LINKED ISSUE #262: Integrate Mozilla's speech api into the gateway frontend

### PR DESCRIPTION
**Pull request by andrenatal**
fixes #262

### CODE REVIEW

**Comment by andrenatal** [comment]
@benfrancis This PR covers the integration of the speech api into the gateway frontend. I want to land this first before we start integrating with the REST intent parser

File: package.json [hunk 1]
```diff
@@ -12,7 +12,8 @@
     "cov": "jest --coverage",
     "start": "webpack && node build/gateway.js",
     "debug-ide": "webpack && node --debug-brk=5858 build/gateway.js",
-    "debug": "webpack && node --inspect build/gateway.js"
+    "debug": "webpack && node --inspect build/gateway.js",
+    "postinstall": "cp node_modules/speaktome-web/build/stm_web.min.js static/js/lib/"
```
Comment (benfrancis) [inline package.json]: Interesting, I had a similar issue with the client side router library and was wondering what to do for dependency management for front end things.

This seems a bit brittle though if the package gets updated. @hobinjk do you know of a better way of doing this?

File: package.json [hunk 2]
```diff
@@ -43,6 +44,7 @@
     "onoff": "^1.1.2",
     "openzwave-shared": "^1.3.2",
     "serialport": "^4.0.7",
+    "speaktome-web": "https://github.com/mozilla/speaktome-web",
```
Comment (benfrancis) [inline package.json]: Are you eventually going to publish this on npm?

File: static/css/app.css [hunk 3]
```diff
@@ -108,6 +108,26 @@ h1, h2, h3, h4, h5, h6 {
   width: 12.5rem;
 }
 
+#speech-button {
+  position: fixed;
```
Comment (benfrancis) [inline static/css/app.css]: I just landed an .icon-button CSS class in app.css which will let you remove most of this CSS if you apply it to the speech-button element.

File: static/js/app.js [hunk 4]
```diff
@@ -41,6 +43,8 @@ var App = {
     this.currentView = 'things';
     this.menuButton = document.getElementById('menu-button');
     this.menuButton.addEventListener('click', Menu.toggle.bind(Menu));
+    this.speechButton = document.getElementById('speech-button');
+    this.speechButton.addEventListener('click', Speech.listen.bind(Speech));
```
Comment (benfrancis) [inline static/js/app.js]: I'd suggest doing this in speech.js

File: static/js/speech.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          this.listening = true;
+        }
+    },
+
+    /**
+     * Listener for the api. Receives a msg containing the current state
+     * @param msg
+     */
+    listener: function(msg) {
+        console.log('Listener msg: ', msg.state);
```
Comment (benfrancis) [inline static/js/speech.js]: Maybe remove this to reduce log noise?

File: static/js/speech.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        if (msg.state == 'result') {
+          // sort results to get the one with the highest confidence
+          var results = msg.data.sort(function(a, b) {
+              a.confidence - b.confidence;
+          });
+          // call the intent rest api with the sentence we decoded
+          console.log(results[0].text, results[0].confidence);
```
Comment (benfrancis) [inline static/js/speech.js]: ...and this?

File: static/js/speech.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        else if (msg.state == 'ready') {
+          this.listening = false;
+          console.log('this.listening' , this.listening);
```
Comment (benfrancis) [inline static/js/speech.js]: ...and this?

**Comment by benfrancis** [comment]
We should put this UI behind a flag in settings which is turned off by default, like floorplan currently is. See https://github.com/mozilla-iot/gateway/pull/254

File: package.json (hunk 2 above)
Comment (andrenatal) [reply package.json]: Yes, we published already. Am fixing this.

Comment (andrenatal) [reply package.json]: The module is here btw https://www.npmjs.com/package/speaktome-api

File: static/js/speech.js (hunk 5 above)
Comment (andrenatal) [reply static/js/speech.js]: fixed

File: static/js/speech.js (hunk 6 above)
Comment (andrenatal) [reply static/js/speech.js]: fixed

File: static/js/speech.js (hunk 7 above)
Comment (andrenatal) [reply static/js/speech.js]: fixed

File: package.json (hunk 1 above)
Comment (andrenatal) [reply package.json]: I implemented a copywebpack plugin here as suggested by @hobinjk

File: static/css/app.css (hunk 3 above)
Comment (andrenatal) [reply static/css/app.css]: Fixed.

File: static/js/app.js (hunk 4 above)
Comment (andrenatal) [reply static/js/app.js]: Fixed.

**Comment by andrenatal** [comment]
@benfrancis All issues you pointed were addressed. Can we land this so I can finish the integration with the intent parser?

File: static/js/settings.js [hunk 8]
```diff
@@ -122,6 +124,29 @@ var SettingsScreen = {
       console.error('Error getting floorplan experiment setting ' + e);
     });
 
+    window.API.getExperimentSetting('speech')
```
Comment (hobinjk) [inline static/js/settings.js]: nitpick: this splits up the getExperimentSetting('floorplan') block above from the logically related `this.floorplanExperimentCheckbox.addEventListener` block below

File: static/js/speech.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+'use strict';
+
+// eslint-disable-next-line no-unused-vars
+var Speech = {
+
+    /**
+    * Initialise menu.
+    */
+    init: function(that) {
```
Comment (hobinjk) [inline static/js/speech.js]: nit: `that` should be called `app` or something that makes it clear what's happening

File: static/js/speech.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Listener for the api. Receives a msg containing the current state
+     * @param msg
+     */
+    listener: function(msg) {
+        if (msg.state == 'result') {
+          // sort results to get the one with the highest confidence
+          var results = msg.data.sort(function(a, b) {
+              a.confidence - b.confidence;
```
Comment (hobinjk) [inline static/js/speech.js]: I think this is missing a `return`, although I believe that the SpeakToMe API results are already sorted by confidence.

File: static/js/settings.js (hunk 8 above)
Comment (andrenatal) [reply static/js/settings.js]: Thanks! Just reordered them!

File: static/js/speech.js (hunk 9 above)
Comment (andrenatal) [reply static/js/speech.js]: Fixed!

File: static/js/speech.js (hunk 10 above)
Comment (andrenatal) [reply static/js/speech.js]: Fixed

**Comment by benfrancis** [comment]
@andrenatal Looks good to me, thank you.

**Comment by andrenatal** [comment]
thank you both @benfrancis @hobinjk . I had all issues addressed and am merging this.

**Pull request merged**